src/lsm/forest.zig

```diff
@@ -162,22 +179,506 @@ pub fn ForestType(comptime _Storage: type, comptime groove_cfg: anytype) type {
+
+            grid: *Grid,
+
+            /// Raw, linear buffer of blocks + reads / writes that will be split up.
+            compaction_blocks: []CompactionHelper.CompactionBlock,
+
+            compactions: stdx.BoundedArray(CompactionInterface, compaction_count) = .{},
+
+            bar_active: CompactionBitset = CompactionBitset.initEmpty(),
+            beat_active: CompactionBitset = CompactionBitset.initEmpty(),
+            beat_reserved: CompactionBitset = CompactionBitset.initEmpty(),
+
+            // TODO: This whole interface around slot_filled_count / slot_running_count needs to be
+            // refactored.
+            slots: [3]?PipelineSlot = .{ null, null, null },
+            slot_filled_count: usize = 0,
+            slot_running_count: usize = 0,
+
+            // Used for invoking the CPU work after a next_tick.
+            // FIXME: Should be a ptr!
+            cpu_slot: ?usize = null,
+
+            state: enum { filling, full, draining, drained } = .filling,
+
+            next_tick: Grid.NextTick = undefined,
+            // next_tick will just invoke our CPU work sync. We need to submit to the underlying
+            // event loop before doing this!
+            // cpu_timeout_completion: IO.Completion = undefined,
+
+            forest: ?*Forest = null,
+            callback: ?*const fn (*Forest) void = null,
+
+            pub fn init(allocator: mem.Allocator, grid: *Grid) !CompactionPipeline {
+                // TODO: Hardcode compaction memory usage for now.
+                // CompactionBlock has some metadata, but we only look at the Block for now which
+                // makes up the bulk.
+                const compaction_memory = 256 * 1024 * 1024; // 256MB.
+                const block_count = @divExact(compaction_memory, constants.block_size);
+                std.log.info("{} blocks for compaction.", .{block_count});
+
+                const compaction_blocks = try allocator.alloc(CompactionHelper.CompactionBlock, block_count);
+                errdefer allocator.free(compaction_blocks);
+
+                for (compaction_blocks, 0..) |*compaction_block, i| {
+                    errdefer for (compaction_blocks[0..i]) |block| allocator.free(block.block);
+                    compaction_block.* = .{
+                        .block = try allocate_block(allocator),
+                    };
+                }
+                errdefer for (compaction_blocks) |block| allocator.free(block.block);
+
+                return .{
+                    .compaction_blocks = compaction_blocks,
+                    .grid = grid,
+                };
+            }
+
+            pub fn deinit(self: *CompactionPipeline, allocator: mem.Allocator) void {
+                for (self.compaction_blocks) |block| allocator.free(block.block);
+                allocator.free(self.compaction_blocks);
+            }
+
+            pub fn reset(self: *CompactionPipeline) void {
```

Comment:
I'm very rusty with the reset code - can't it be called at any arbitrary moment?

Reply:
Ah, you are actually correct! We might call reset with in-progress compaction, and that's actually the reason why we need reset at all!

The situation is like this:

* a replica compacts
* it gets stuck reading a block due to local corruption
* and the block is missing from the entire cluster, because the rest of the replicas moved over to the next checkpoint. 

in his situation:

* replica gets a newer checkpoint
* replica _cancels_ all outstanding grid reads and writes (dropping callbacks on the floor)
* replica installs the new superblock
* replica resets memory of all pending data structurues

So, yeah, this absolutely can get reset while some compaction is stuck on reading